logstash-core/lib/logstash/config/mixin.rb

```diff
@@ -126,6 +146,42 @@ def config_init(params)
     @config = params
   end # def config_init
 
+  # Replace all environment variable references in 'value' param by environment variable value and return updated value
+  # Process following patterns : $VAR, ${VAR}, ${VAR:defaultValue}
+  def replace_env_placeholders(value)
+    if (value.is_a?(String))
+      while value =~ ENV_PLACEHOLDER_REGEX
```

Comment:
I think there's a shorter/simpler way to do this. Doing it shorter will require two things:

1) Using named captures in your ENV_PLACEHOLDER_REGEX
2) Using `String#gsub`

Rough proposal (I haven't tested thoroughly). Demonstrated in irb:

```ruby
>> ENV_PLACEHOLDER_REGEX = Regexp.new(/\$(?<env>\w+)|\$\{(?<env>\w+)(\:(?<default>[^}]*))?\}/)
>> text = "hello ${world} ${fancy} whoa ${missing:cool default}"
=> "hello ${world} ${fancy} whoa ${missing:cool default}"

# Demonstrating it using String#gsub and regexp named captures
>> text.gsub(ENV_PLACEHOLDER_REGEX) { |t| m = Regexp.last_match; value = map[m["env"]]; value.nil? ? m["default"] : value }
=> "hello earth pants whoa cool default"
```

Using named captures will allow you to lookup the env var name requested instead of having to compute it (as in lines 154-160) and also let you easily look up the default value (if provided).

---

I am not sure I am making sense here, so let me know if I'm being confusing. My overall feeling is that moving to use String#gsub + named captures allows you to shrink lines 152-180 to be something much more concise (and hopefully easier to debug should things go wrong)